presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java

```diff
@@ -178,9 +178,22 @@ public TestingPrestoServer(boolean coordinator,
             SqlParserOptions parserOptions,
             List<Module> additionalModules)
             throws Exception
+    {
+        this(coordinator, properties, environment, discoveryUri, parserOptions, additionalModules, Optional.empty());
+    }
+
+    public TestingPrestoServer(boolean coordinator,
+            Map<String, String> properties,
+            String environment,
+            URI discoveryUri,
+            SqlParserOptions parserOptions,
+            List<Module> additionalModules,
+            Optional<Path> baseDataDirPath)
+            throws Exception
     {
         this.coordinator = coordinator;
-        baseDataDir = Files.createTempDirectory("PrestoTest");
+
+        baseDataDir = baseDataDirPath.orElse(Files.createTempDirectory("PrestoTest"));
```

Comment:
you should use the lambda version of `orElse` to avoid the creation of the temp directory.  Also I'd static import that method.

I'd also perform some basic validation on the path since it is now user supplied.  If it already exists, it should be a directory.  If not, either throw an exception or possibly create the directory (not sure which is more desirable).

Reply:
The cleanup code also needs to be changed not to delete the directory if it is user provided.

Reply:
@dain The validation was done in RawLocalFileSystem.mkdirsWithOptionalPermission() and ChecksumFileSystem.create() when creating database. In that case exceptions would be thrown on invalid user supplied path when creating database. I didn't include the validation in HiveQueryRunner.main() originally because I thought the logic would be repeated twice. But doing it in HiveQueryRunner would allow invalid path to fail sooner than later in createDatabase(), so I copied the validation logic to HiveQueryRunner.main() in the new commit.

> I'd also perform some basic validation on the path since it is now user supplied. If it already exists, it should be a directory. If not, either throw an exception or possibly create the directory (not sure which is more desirable).